Publish OSDK and OSTD dry-run fails when OSTD depends on unpublished local ostd-macros changes

## Bug Report

The `Publish OSDK and OSTD` workflow currently fails in the dry-run publish step when packaging `ostd`.

Recent main-branch runs of this workflow have also failed, for example:

- https://github.com/asterinas/asterinas/actions/runs/28701482349
- https://github.com/asterinas/asterinas/actions/runs/27815745299
- https://github.com/asterinas/asterinas/actions/runs/27815248905

## Error

The failure happens while verifying the packaged `ostd` crate:

```text
error[E0432]: unresolved import `ostd_macros::early_cmdline_parser`
  --> src/lib.rs:60:5
   |
60 |     early_cmdline_parser, global_frame_allocator, global_heap_allocator,
   |     ^^^^^^^^^^^^^^^^^^^^ no `early_cmdline_parser` in the root

error: could not compile `ostd` (lib) due to 1 previous error
error: failed to verify package tarball
```

## Root Cause

`ostd/libs/ostd-macros/src/lib.rs` in the current workspace does define and export `early_cmdline_parser`.

However, during `cargo publish --dry-run` for `ostd`, Cargo verifies the packaged crate as if it were published. The local path dependency on `ostd-macros` is resolved as a registry dependency, so Cargo downloads the already-published `ostd-macros v0.18.0` from crates.io.

That published version does not contain `early_cmdline_parser`, while the current local `ostd` source imports it. As a result, the dry-run verification fails.

## Impact

This can cause PR CI to fail in the `publish` job even when the PR does not touch publishing logic or `ostd-macros`.

For example, a PR that only changes `ostd/src/smp.rs` can still hit this failure because the publish dry-run validates the current workspace's `ostd` crate against crates.io's older `ostd-macros` package.

## Possible Fixes

One of the following approaches may be needed:

- Adjust `tools/github_workflows/publish_osdk_and_ostd.sh` so PR dry-runs for interdependent workspace crates do not verify against stale crates.io versions.
- Use `cargo publish --dry-run --no-verify` or `cargo package` for crates whose dependencies are unpublished workspace crates.
- Ensure dependent workspace crates such as `ostd-macros` are published before verifying crates that depend on newly added macro exports.

The core issue is that dry-run verification does not use the just-built local `ostd-macros` package when validating `ostd`.

--- a/.github/workflows/publish_osdk_and_ostd.yml
+++ b/.github/workflows/publish_osdk_and_ostd.yml
@@ -7,6 +7,8 @@ on:
       - VERSION
       - ostd/**
       - osdk/**
+      - .github/workflows/publish_osdk_and_ostd.yml
+      - tools/github_workflows/publish_osdk_and_ostd.sh
   push:
     branches:
       - main

--- a/tools/github_workflows/publish_osdk_and_ostd.sh
+++ b/tools/github_workflows/publish_osdk_and_ostd.sh
@@ -41,24 +41,90 @@ while [ "$#" -gt 0 ]; do
     esac
 done
 
-# Performs the publish check or publish the crate in directory $1.
-# All the arguments after $1 are passed to `cargo publish`.
-do_publish_for() {
-    pushd $ASTER_SRC_DIR/$1
+# Tracks packages that have already passed `cargo publish --dry-run` in this
+# script. `--dry-run` does not really publish earlier packages to `crates.io`,
+# so later packages patch only these already-checked local `path` dependencies
+# to mirror the dependency state that a real publish would see.
+declare -A DRY_RUN_PUBLISHED_PACKAGES=()
 
-    ADDITIONAL_ARGS="${@:2}"
-    RF="$RUSTFLAGS --check-cfg cfg(ktest)"
+mark_dry_run_published() {
+    DRY_RUN_PUBLISHED_PACKAGES[$1]=true
+}
+
+is_dry_run_published() {
+    [ -n "${DRY_RUN_PUBLISHED_PACKAGES[$1]}" ]
+}
+
+# Performs the publish check or publishes the crate in directory $1.
+# All arguments after $1 are passed to `cargo publish`.
+do_publish_for() {
+    pushd "$ASTER_SRC_DIR/$1"
+    shift
+
+    local ADDITIONAL_ARGS=("$@")
+    local RF="$RUSTFLAGS --check-cfg cfg(ktest)"
 
     if [ -n "$DRY_RUN" ]; then
+        local DEP_NAME DEP_PATH LOCAL_PATH_DEPS MANIFEST_PATH METADATA PACKAGE_METADATA PACKAGE_NAME
+        local PATCH_ARGS=()
+
+        MANIFEST_PATH="$(pwd -P)/Cargo.toml"
+        # `cargo metadata` prints a JSON document for the **whole workspace**,
+        # not only for `MANIFEST_PATH`. Each package entry includes fields such as
+        # `manifest_path`, `name`, and `dependencies`; the `--dry-run` logic below
+        # uses them to find the package being published and its local `path` deps.
+        METADATA=$(cargo metadata --format-version=1 --no-deps --manifest-path "$MANIFEST_PATH")
+
+        # Pick the package entry for the manifest currently being published.
+        PACKAGE_METADATA=$(jq --arg manifest "$MANIFEST_PATH" '
+            .packages[]
+            | select(.manifest_path == $manifest)
+        ' <<< "$METADATA")
+        if [ -z "$PACKAGE_METADATA" ]; then
+            echo "Error: cargo metadata did not contain a package for manifest: $MANIFEST_PATH" >&2
+            exit 1
+        fi
+
+        PACKAGE_NAME=$(jq -r '.name' <<< "$PACKAGE_METADATA")
+        if [ -z "$PACKAGE_NAME" ]; then
+            echo "Error: cargo metadata did not contain a package name for manifest: $MANIFEST_PATH" >&2
+            exit 1
+        fi
+
+        # List local non-dev `path` dependencies as tab-separated
+        # `name<TAB>path` rows.
+        LOCAL_PATH_DEPS=$(jq -r '
+            .dependencies[]
+            | select(.path != null and .source == null and .kind != "dev")
+            | [.name, .path]
+            | @tsv
+        ' <<< "$PACKAGE_METADATA")
+
+        if [ -n "$LOCAL_PATH_DEPS" ]; then
+            while IFS=$'\t' read -r DEP_NAME DEP_PATH; do
+                if is_dry_run_published "$DEP_NAME"; then
+                    PATCH_ARGS+=(--config "patch.crates-io.$DEP_NAME.path=\"$DEP_PATH\"")
+                fi
+            done <<< "$LOCAL_PATH_DEPS"
+        fi
+
         # Perform checks
-        RUSTFLAGS=$RF cargo publish --dry-run --allow-dirty $ADDITIONAL_ARGS
-        RUSTFLAGS=$RF cargo doc $ADDITIONAL_ARGS
+        RUSTFLAGS=$RF cargo publish --dry-run --allow-dirty "${PATCH_ARGS[@]}" "${ADDITIONAL_ARGS[@]}"
+        RUSTFLAGS=$RF cargo doc "${ADDITIONAL_ARGS[@]}"
+
+        mark_dry_run_published "$PACKAGE_NAME"
     else
-        RUSTFLAGS=$RF cargo publish --token $TOKEN $ADDITIONAL_ARGS
+        RUSTFLAGS=$RF cargo publish --token $TOKEN "${ADDITIONAL_ARGS[@]}"
     fi
 
     popd
 }
+
+# The order of `do_publish_for` calls matters:
+# `do_publish_for path/to/crate-a`
+# must appear before
+# `do_publish_for path/to/crate-b`
+# if `crate-a` is a dependency of `crate-b`.
 
 do_publish_for ostd/libs/linux-bzimage/boot-params
 do_publish_for ostd/libs/linux-bzimage/builder

